Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.cpp

```diff
@@ -41,32 +41,77 @@
 namespace WebKit {
 using namespace WebCore;
 
+
+class RemoteImageBufferProxyFlushFence : public ThreadSafeRefCounted<RemoteImageBufferProxyFlushFence> {
+    WTF_MAKE_NONCOPYABLE(RemoteImageBufferProxyFlushFence);
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static Ref<RemoteImageBufferProxyFlushFence> create(IPC::Semaphore semaphore)
+    {
+        return adoptRef(*new RemoteImageBufferProxyFlushFence { WTFMove(semaphore) });
+    }
+
+    ~RemoteImageBufferProxyFlushFence()
+    {
+        if (!m_signaled)
+            tracePoint(FlushRemoteImageBufferEnd, reinterpret_cast<uintptr_t>(this), 2u);
+    }
+
+    bool waitFor(Seconds timeout)
+    {
+        Locker locker { m_lock };
+        if (m_signaled)
+            return true;
+        m_signaled = m_semaphore.waitFor(timeout);
+        if (m_signaled)
+            tracePoint(FlushRemoteImageBufferEnd, reinterpret_cast<uintptr_t>(this), 0u);
+        return m_signaled;
+    }
+
+    void forceSignal()
```

Comment:
Is there any value in this function?

If a background thread is already waiting, then it'll block until that thread releases the lock (3 seconds at most), at which point we no longer need to signal the semaphore.

Could we just not bother? The only caller immediately drops its ref to this object afterwards.

I guess there's a chance that the background thread will start waiting after this call, but that seems like a race, where the current behaviour is we choose to block the main thread unnecessarily, and the alternative is to block the background thread unnecessarily.